src/ui/localization/translator.ts

```diff
@@ -0,0 +1,246 @@
+import en from "./locale/en.json";
+import es from "./locale/es.json";
+import it from "./locale/it.json";
+import type { PeriodUnit } from "../../helpers/duration-helper";
+
+/**
+ * Custom translations to be used in the purchase flow.
+ * This class allows you to override the default translations used in the purchase flow.
+ * The main level keys are the locale codes and the values are objects with the same keys as the default translations.
+ * @public
+ * @example
+ * This example will override the default translation for the email step title in the English locale.
+ * ```typescript
+ * const customTranslations = {
+ *  en: {
+ *    "state_needs_auth_info.email_step_title": "Billing email",
+ *   }
+ * }
+ * ```
+ */
+export type CustomTranslations = Record<string, Record<string, string>>;
+
+/**
+ * Translation variables to be used in the translation.
+ * This class allows you to pass variables to the translate method.
+ * @public
+ * @example Given a label with id `periods.monthPlural` and value `{{amount}} months`. This example will replace the variable `{{amount}}` with the value `10`.
+ * ```typescript
+ * translator.translate('periods.monthPlural', { amount: 10 });
+ * // Output: 10 months
+ * ```
+ */
+export type TranslationVariables = Record<
+  string,
+  string | number | undefined | null
+>;
+
+export interface TranslatePeriodOptions {
+  noWhitespace?: boolean;
+  short?: boolean;
+}
+
+const defaultTranslatePeriodOptions: TranslatePeriodOptions = {
+  noWhitespace: false,
+  short: false,
+};
+
+export class Translator {
+  public readonly locales: Record<string, LocaleTranslations> = {};
+
+  public static fallback() {
+    return new Translator();
+  }
+
+  public constructor(
+    customTranslations: CustomTranslations = {},
+    public readonly selectedLocale: string = "en",
+    public readonly defaultLocale: string = "en",
+  ) {
+    this.locales = {
+      en: new LocaleTranslations(en, "en"),
+      es: new LocaleTranslations(es, "es"),
+      it: new LocaleTranslations(it, "it"),
```

Comment:
Nitpick but should we have a `supportedLocales` array and create this dictionary dynamically, so it's easier to add new languages? We could also then have a test to make sure all supported locales have corresponding localization jsons.

Reply:
Bump to this one. Small cleanup but just wondering if we can build this dynamically, and maybe even have a test that checks that all locales in the `locale` folder are added to the `SUPPORTED_LOCALES` constant. This way we won't forget to add new locales here.

Reply:
> This way we won't forget to add new locales here.

Yeah, which is already the case given that @MrEdwardo added more locales 😅 

I'll create a reference list for now, I would avoid "scanning folders" or something similar just to make sure that we still bundle all these in the UMD and we don't rely on them being hosted somewhere.

Reply:
I moved the supported languages in a dedicated file and initialized the `locales` attribute here using that object.

I wouldn't make this dynamic with the included files since this has to by resolved at build time.

I added a test to check that every file that is present in the `/locale` folder gets imported in the `supportedLanguages` object.

Reply:
Yeah agreed. I think adding a test is enough. Thanks!